File: omp/src20/solver_PCG.c
/*
 * solver_PCG
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <math.h>
#include <omp.h>

#include "solver_PCG.h"

extern int
solve_PCG (int N, int NL, int NU, int *indexL, int *itemL, int *indexU, int *itemU,
		double *D, double *B, double *X, double *AL, double *AU,
  	        double EPS, int *ITR, int *IER, int N2)
{
	double **W;
	double VAL, BNRM2, WVAL, SW, RHO, BETA, RHO1, C1, DNRM2, ALPHA, ERR;
	double Stime, Etime;
	int i, j, ic, ip, L, ip1, N3;
	int R = 0;
	int Z = 1;
	int Q = 1;
	int P = 2;
	int DD = 3;

/*********
 * INIT. *
 *********/
        N3= N;
	W = (double **)malloc(sizeof(double *)*4);
	if(W == NULL) {
		fprintf(stderr, "Error: %s\n", strerror(errno));
		return -1;
	}
	for(i=0; i<4; i++) {
  	    W[i] = (double *)malloc(sizeof(double)*N3);
		if(W[i] == NULL) {
			fprintf(stderr, "Error: %s\n", strerror(errno));
			return -1;
		}
	}

#pragma omp parallel for private (i)
	for(i=0; i<N; i++) {
		X[i] = 0.0;
		W[1][i] = 0.0;
		W[2][i] = 0.0;
		W[3][i] = 0.0;
	}

/**************************
 * {r0} = {b} - {A}{xini} *
 **************************/
#pragma omp parallel for private (i,VAL,j)
	for(i=0; i<N; i++) {
		VAL = D[i] * X[i];
		for(j=indexL[i]; j<indexL[i+1]; j++) {
			VAL += AL[j] * X[itemL[j]-1];
		}
		for(j=indexU[i]; j<indexU[i+1]; j++) {
			VAL += AU[j] * X[itemU[j]-1];
		}
		W[R][i] = B[i] - VAL;
	}

	BNRM2 = 0.0;
#pragma omp parallel for private (i) reduction (+:BNRM2)
	for(i=0; i<N; i++) {
	  BNRM2 += B[i]*B[i];
	}

#pragma omp parallel for private (i)
	for(i=0; i<N; i++) {
	  W[DD][i]= 1.e0/D[i];
	}

/************************************************************** ITERATION */
	*ITR = N;

	Stime = omp_get_wtime();
	for(L=0; L<(*ITR); L++) {

/*******************
 * {z} = [Minv]{r} *
 *******************/
#pragma omp parallel for private (i)
	  for(i=0; i<N; i++) {
	    W[Z][i] = W[R][i]*W[DD][i];
	  }
/****************
 * RHO = {r}{z} *
 ****************/
		RHO = 0.0;
#pragma omp parallel for private (i) reduction(+:RHO)
		for(i=0; i<N; i++) {
		  RHO += W[R][i] * W[Z][i];
		}

/********************************
 * {p}  = {z} if      ITER=0    *
 * BETA = RHO / RHO1  otherwise *
 ********************************/
		if(L == 0) {
#pragma omp parallel for private (i)
		  for(i=0; i<N; i++) {
			W[P][i] = W[Z][i];
		  }
		} else {
		  BETA = RHO / RHO1;
#pragma omp parallel for private (i)
		  for(i=0; i<N; i++) {
			W[P][i] = W[Z][i] + BETA * W[P][i];
		}
		}

/****************
 * {q} = [A]{p} *
 ****************/
#pragma omp parallel for private (i,VAL,j)
		for(i=0; i<N; i++) {
		  VAL = D[i] * W[P][i];
		  for(j=indexL[i]; j<indexL[i+1]; j++) {
			VAL += AL[j] * W[P][itemL[j]-1];
		  }
		  for(j=indexU[i]; j<indexU[i+1]; j++) {
			VAL += AU[j] * W[P][itemU[j]-1];
		  }
		  W[Q][i] = VAL;
		}

/************************
 * ALPHA = RHO / {p}{q} *
 ************************/
		C1 = 0.0;
#pragma omp parallel for private (i) reduction(+:C1)
		for(i=0; i<N; i++) {
			C1 += W[P][i] * W[Q][i];
		}

		ALPHA = RHO / C1;

/***************************
 * {x} = {x} + ALPHA * {p} *
 * {r} = {r} - ALPHA * {q} *
 ***************************/
#pragma omp parallel for private (i)
		for(i=0; i<N; i++) {
			X[i]    += ALPHA * W[P][i];
			W[R][i] -= ALPHA * W[Q][i];
		}

		DNRM2 = 0.0;
#pragma omp parallel for private (i) reduction(+:DNRM2)
		for(i=0; i<N; i++) {
		  DNRM2 += W[R][i]*W[R][i];
		}

		ERR = sqrt(DNRM2/BNRM2);
                if( (L+1)%100 ==1) {
                        fprintf(stderr, "%5d%16.6e\n", L+1, ERR);
                }

		if(ERR < EPS) {
			*IER = 0;
			goto N900;
		} else {
			RHO1 = RHO;
		}
	}
	*IER = 1;

N900:
	Etime = omp_get_wtime();
	fprintf(stderr, "%5d%16.6e\n", L+1, ERR);
	fprintf(stderr, "%16.6e sec. (solver)\n", Etime - Stime);
	*ITR = L;

	free(W);

	return 0;
}

File: multicore-f/L1/src/solver_PCG.f
!C*** 
!C*** module solver_PCG
!C***
!
      module solver_PCG
      contains
!C
!C*** solve_PCG
!C
      subroutine solve_PCG                                              &
     &         ( N, NPL, NPU, indexL, itemL, indexU, itemU, D, B, X,    &
     &           AL, AU, EPS, ITR, IER)

      implicit REAL*8 (A-H,O-Z)

      real(kind=8), dimension(N)   :: D
      real(kind=8), dimension(N)   :: B
      real(kind=8), dimension(N)   :: X
      real(kind=8), dimension(NPL) :: AL
      real(kind=8), dimension(NPU) :: AU

      integer, dimension(0:N)   :: indexL, indexU
      integer, dimension(NPL):: itemL
      integer, dimension(NPU):: itemU

      real(kind=8), dimension(:,:), allocatable :: W

      integer, parameter ::  R= 1
      integer, parameter ::  Z= 2
      integer, parameter ::  Q= 2
      integer, parameter ::  P= 3
      integer, parameter :: DD= 4

!C
!C +------+
!C | INIT |
!C +------+
!C===
      allocate (W(N,4))

      do i= 1, N
        X(i)  = 0.d0
        W(i,1)= 0.0D0
        W(i,2)= 0.0D0
        W(i,3)= 0.0D0
        W(i,4)= 0.0D0
      enddo

      do i= 1, N
        W(i,DD)= 1.d0/D(i)
      enddo
!C===

!C
!C +-----------------------+
!C | {r0}= {b} - [A]{xini} |
!C +-----------------------+
!C===
      do i= 1, N
        VAL= D(i)*X(i)
        do k= indexL(i-1)+1, indexL(i)
          VAL= VAL + AL(k)*X(itemL(k))
        enddo 
        do k= indexU(i-1)+1, indexU(i)
          VAL= VAL + AU(k)*X(itemU(k))
        enddo 
        W(i,R)= B(i) - VAL
      enddo

      BNRM2= 0.0D0
      do i= 1, N
        BNRM2 = BNRM2 + B(i)  **2
      enddo
!C===

!C
!C***************************************************************  ITERATION
      ITR= N

      do L= 1, ITR
!C
!C +----------------+
!C | {z}= [Minv]{r} |
!C +----------------+
!C===      
      do i= 1, N
        W(i,Z)= W(i,R)*W(i,DD)
      enddo
!C===

!C
!C +-------------+
!C | RHO= {r}{z} |
!C +-------------+
!C===
      RHO= 0.d0
      do i= 1, N
        RHO= RHO + W(i,R)*W(i,Z)   
      enddo          
!C===

!C
!C +-----------------------------+
!C | {p} = {z} if      ITER=1    |
!C | BETA= RHO / RHO1  otherwise |
!C +-----------------------------+
!C===
      if ( L.eq.1 ) then

        do i= 1, N
          W(i,P)= W(i,Z)
        enddo
       else
         BETA= RHO / RHO1
         do i= 1, N
           W(i,P)= W(i,Z) + BETA*W(i,P)
         enddo
      endif
!C===

!C
!C +-------------+
!C | {q}= [A]{p} |
!C +-------------+
!C===        
      do i= 1, N
        VAL= D(i)*W(i,P)
        do k= indexL(i-1)+1, indexL(i)
          VAL= VAL + AL(k)*W(itemL(k),P)
        enddo 
        do k= indexU(i-1)+1, indexU(i)
          VAL= VAL + AU(k)*W(itemU(k),P)
        enddo 
        W(i,Q)= VAL
      enddo
!C===

!C
!C +---------------------+
!C | ALPHA= RHO / {p}{q} |
!C +---------------------+
!C===
      C1= 0.d0
      do i= 1, N
        C1= C1 + W(i,P)*W(i,Q)
      enddo

      ALPHA= RHO / C1
!C===


!C
!C +----------------------+
!C | {x}= {x} + ALPHA*{p} |
!C | {r}= {r} - ALPHA*{q} |
!C +----------------------+
!C===
      do i= 1, N
        X(i)  = X(i)   + ALPHA * W(i,P)
        W(i,R)= W(i,R) - ALPHA * W(i,Q)
      enddo

      DNRM2= 0.d0
      do i= 1, N
        DNRM2= DNRM2 + W(i,R)**2
      enddo
!C===

      ERR = dsqrt(DNRM2/BNRM2)
      if (mod(L,100).eq.1) then
        write (*,'(i5,2(1pe16.6))') L, ERR
      endif

        if (ERR .lt. EPS) then
          IER = 0
          goto 900
         else
          RHO1 = RHO
        endif

      enddo
      IER = 1

  900 continue

      write (*,'(i5,2(1pe16.6))') L, ERR
      ITR= L
!      EPS= ERR

      deallocate (W)

      return

      end subroutine  solve_PCG
      end module     solver_PCG



